keps/sig-network/3766-referencegrant/README.md

```diff
@@ -0,0 +1,492 @@
+support referencing the To referent, and must reference the one in the To field.
+
+With all of that in mind, we felt that the benefits in terms of specificity
+outweighed the costs in terms of user pain to use the more correct `Resource`.
+
+### API Spec
+
+```golang
+// ReferenceGrant identifies kinds of resources in other namespaces that are
+// trusted to reference the specified kinds of resources in the same namespace
+// as the policy.
+//
+// Each ReferenceGrant can be used to represent a unique trust relationship.
+// Additional ReferenceGrants can be used to add to the set of trusted
+// sources of inbound references for the namespace they are defined within.
+//
+// ReferenceGrant is a form of runtime verification allowing users to assert
+// which cross-namespace object references are permitted. Implementations that
+// support ReferenceGrant MUST NOT permit cross-namespace references which have
+// no grant, and MUST respond to the removal of a grant by revoking the access
+// that the grant allowed.
+//
+// Support: Core
+type ReferenceGrant struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	// Spec defines the desired state of ReferenceGrant.
+	Spec ReferenceGrantSpec `json:"spec,omitempty"`
+
+	// Note that `Status` sub-resource has been excluded at the
+	// moment as it was difficult to work out the design.
+	// A `Status` sub-resource may be added in the future.
+}
+
+// +kubebuilder:object:root=true
+// ReferenceGrantList contains a list of ReferenceGrant.
+type ReferenceGrantList struct {
+	metav1.TypeMeta `json:",inline"`
+	metav1.ListMeta `json:"metadata,omitempty"`
+	Items           []ReferenceGrant `json:"items"`
+}
+
+// ReferenceGrantSpec identifies a cross namespace relationship that is trusted
+// for Gateway API.
+type ReferenceGrantSpec struct {
+	// From describes the trusted namespaces and kinds that can reference the
+	// resources described in "To". Each entry in this list MUST be considered
+	// to be an additional place that references can be valid from, or to put
+	// this another way, entries MUST be combined using OR.
+	//
+	// +kubebuilder:validation:MinItems=1
+	// +kubebuilder:validation:MaxItems=16
+	From []ReferenceGrantFrom `json:"from"`
+
+	// To describes the resources that may be referenced by the resources
+	// described in "From". Each entry in this list MUST be considered to be an
+	// additional place that references can be valid to, or to put this another
+	// way, entries MUST be combined using OR.
+	//
+	// +kubebuilder:validation:MinItems=1
+	// +kubebuilder:validation:MaxItems=16
+	To []ReferenceGrantTo `json:"to"`
+}
+
+// ReferenceGrantFrom describes trusted namespaces and kinds.
+type ReferenceGrantFrom struct {
+	// Group is the group of the referent.
+	// When empty, the Kubernetes core API group is inferred.
+	Group Group `json:"group"`
+
+	// Kind is the kind of the referent.
+	Kind string `json:"kind"`
```

Comment:
hoisting this question from the [sig-auth discussion](https://docs.google.com/document/d/1woLGRoONE3EBVx-wTb4pvp4CI7tmLZ6lS26VTbosLKM/edit#bookmark=id.9q15hi5lip46):

If the same kind can be referenced in two different ways from another API, is group/kind granularity sufficient for the `from` allowlist? As a made up example, if Volumes can reference VolumeSnapshots as a source (readonly) or a destination (write), could I allow my snapshots to be used as a source but not a write destination?

Reply:
Most of the comments in this PR are about this topic :)

Reply:
I agree that it's conceivable that we might want to add extra dimensions to the `from` stanza to further restrict the access in the future, but to me, that sounds like a great place where we could add an optional field in a compatible way later, if we see a demand for it.

I think that the design of any additional dimensions would be pretty tricky too - we'd almost need an annotation-style map[string]string to be able to describe arbitrary access subdivisions. Which give me the yucks, personallly.

Reply:
Taking the volume / volumesnapshot thought experiment further and thinking about safe evolution of references, is it a good idea to leave an indication of the purpose for the reference optional?

Imagine:
1. volumes add the ability to provision from snapshots (read only), so someone creates a ReferenceGrant object to allow referencing all the volumesnapshots in their `templates` namespace with something like `from: [{resource:volumes, group:storage.k8s.io, namespace:*}]` and `to: {resource:volumesnapshots, group:snapshot.storage.k8s.io}`
2. at a later point in time, volumes add the ability to back up to snapshots, qualified by `for: backup` or something in the `from` entries

Would the existing ReferenceGrant that did not scope the purpose for the grant (but assumed it was limited to read-only use, since that was the only use defined at the time of creation) automatically expand to allow any volume in any namespace to reference the snapshots as a write destination for a backup?

Reply:
I think the most straightforward approach would be for this to only grant read only referential access initially. In the future we could add a new optional field that allowed users to grant write access if such a need ever came up.

On the other hand, one of the current uses, forwarding traffic to a Service in another namespace does not clearly fall in the "read" or "write" category to me. I think I could be OK with a `for` field that allowed for a variety of predefined and/or domain-prefixed values. Since this is already within a list, I'd prefer to keep this as an individual string so we don't end up with a list within a list.

Reply:
Introducing a `verbs` or `for` field could actually enable us to relatively safely allow `Any` or `*` as a value for resources or groups in `From`. For the sake of consistency and flexibility, `verbs` seems like it could be a good fit. Maybe we could define some well known/standard values as part of this API but also allow domain-prefixed verbs for other uses of the API?

Reply:
Lot's of comments that are now "outdated" which need followup - Github makes this hard.

I agree that read-only is simpler (and I'd include Service forwarding in that), but I *do* think we should ask the larger question of where we want this to evolve.   At one end is a simple resource which only models read-only access.  At the other end is a robust mechanism which governs cross-ns (and in fact intra-ns, too!) references the same way other authz is done...

This is a sig-auth KEP, so they need to own that analysis.   I don't want to bite off more than we can chew, but...

Reply:
Moved to "open questions" since I think we're going to need more discussion here.